pkg/processor/trigger/kafka/trigger.go

```diff
@@ -228,133 +220,56 @@ func (k *kafka) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.C
-						CustomHandler: nil,
-					})
-				var wg sync.WaitGroup
-				wg.Add(2)
-				go func() {
-					err := <-submittedEventInstance.done
-
-					// we successfully submitted the message to the handler. mark it
-					if err == nil {
-						session.MarkOffset(
-							message.Topic,
-							message.Partition,
-							message.Offset+1-ackWindowSize,
-							"",
-						)
-					}
-					k.Logger.DebugWith("Handler done", "partition", claim.Partition())
-					wg.Done()
-				}()
-				go func() {
-
-					// this needs to occur once. the reason is that this specific function (ConsumeClaim)
-					// runs in parallel for each partition, and we want to make sure that we only
-					// drain the workers once.
-					if err := k.SignalWorkerDraining(); err != nil {
-						k.Logger.DebugWith("Failed to signal worker draining",
-							"err", err.Error(),
-							"partition", claim.Partition())
-					} else {
-						drainedWorker = true
-					}
-					wg.Done()
-				}()
-
-				wg.Wait()
-				readyForRebalanceChan <- true
-			}()
+			// handle in the goroutine so we don't block
+			submittedEventChan <- &submittedEventInstance
 
-			// wait a for rebalance readiness or max timeout
 			select {
-			case <-readyForRebalanceChan:
-				k.Logger.DebugWith("Handler done, rebalancing will commence",
-					"partition", claim.Partition())
-
-			case <-time.After(k.configuration.maxWaitHandlerDuringRebalance):
-				k.Logger.DebugWith("Timed out waiting for handler to complete",
-					"partition", claim.Partition())
-
-				// mark this as a failure, metric-wise
-				k.UpdateStatistics(false)
-
-				// restart the worker, and having failed that shut down
-				if err := k.cancelEventHandling(workerInstance, claim); err != nil {
-					k.Logger.DebugWith("Failed to cancel event handling",
-						"err", err.Error(),
-						"partition", claim.Partition())
-
-					panic("Failed to cancel event handling")
+			case err := <-submittedEventInstance.done:
+
+				// we successfully submitted the message to the handler. mark it
+				if err == nil {
+					session.MarkOffset(
+						message.Topic,
+						message.Partition,
+						message.Offset+1-ackWindowSize,
+						"",
+					)
+				}
+				if err := k.partitionWorkerAllocator.ReleaseWorker(cookie, workerInstance); err != nil {
+					return errors.Wrap(err, "Failed to release worker")
+				}
+			case <-session.Context().Done():
+				k.drainOnRebalance(session, claim, workerInstance, &submittedEventInstance, message, true)
+				if err := k.partitionWorkerAllocator.ReleaseWorker(cookie, workerInstance); err != nil {
+					return errors.Wrap(err, "Failed to release worker")
 				}
+				break loop
 			}
-		}
 
-		// release the worker from whence it came
-		if err := k.partitionWorkerAllocator.ReleaseWorker(cookie, workerInstance); err != nil {
-			return errors.Wrap(err, "Failed to release worker")
+		case <-session.Context().Done():
+			k.drainOnRebalance(session, claim, nil, nil, nil, false)
```

Comment:
Same as above but explain the `false`.

It might seem trivial, but this code is fairly complex as it is and we want to make it easier for future developers to understand it.